### PR TITLE
task 4

### PR DESCRIPTION
**Pull request by NataliaDymnikova**

### CODE REVIEW

File: assignment4/src/CountingIslands.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Class for counting islands in the 2-dimensial array of booleans
+ */
+public class CountingIslands {
+    /**
+     * Count islands
+     *
+     * @param numRows     -- number of rows
+     * @param numCols     -- number of columns
+     * @param isLandArray -- 2-dimensial array, where True is land, False is water
```
Comment (simon-frankau) [inline assignment4/src/CountingIslands.java]: nit: dimensional

File: assignment4/src/CountingIslands.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    private static void deleteIsland(int startRow, int startCol, int numRows, int numCols, boolean[][] isLandArray) {
+        if (!isLandArray[startRow][startCol]) {
+            return;
+        }
+
+        isLandArray[startRow][startCol] = false;
+
+        int minNeighborRow = startRow == 0 ? 1 : startRow - 1;
+        int maxNeighborRow = min(startRow + 1, numRows - 1);
```
Comment (simon-frankau) [inline assignment4/src/CountingIslands.java]: Why use a ternary operator in one case, and "min" in the other? You should be able to take a consistent approach.

File: assignment4/src/CountingIslands.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        for (int row = minNeighborRow; row <= maxNeighborRow; row += 2) {
+            deleteIsland(row, startCol, numRows, numCols, isLandArray);
+        }
+
+        int minNeighborCol = startCol == 0 ? 1 : startCol - 1;
+        int maxNeighborCol = min(startCol + 1, numCols - 1);
+        for (int col = minNeighborCol; col <= maxNeighborCol; col += 2) {
+            deleteIsland(startRow, col, numRows, numCols, isLandArray);
+        }
```
Comment (simon-frankau) [inline assignment4/src/CountingIslands.java]: While two "for" loops do the job, this could just be reduced to 4 explicit recursive calls, which I think would be simpler code.

Also, bear in mind that your approach of recursion, rather than e.g. keeping a queue or stack data structure, could cause stack overflow on very large inputs. For this situation, it's probably fine, but it would be good to have a comment about the algorithm's limitations.

File: assignment4/src/CountingIslandsTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Test
+    public void testIsSomeIslands() {
+        boolean[][] isLands = new boolean[][]{
+                {false, true, false, true},
+                {true, true, false, false},
+                {false, false, true, false},
+                {false, false, true, false}};
+        int islands = CountingIslands.countingIslands(4, 4, isLands);
+        assertEquals(3, islands);
```
Comment (simon-frankau) [inline assignment4/src/CountingIslandsTest.java:42]: Nice test, demonstrating that diagonal connections don't count.

File: assignment4/src/CountingIslandsTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+}
```
Comment (simon-frankau) [inline assignment4/src/CountingIslandsTest.java]: I would liked to have seen a few more tests to do with e.g. islands with lakes in them - especially if the lake contains another island. These are the kinds of corner cases that are easy to get wrong.

**Review by simon-frankau — APPROVED** [review]
Good, simple solution. I'd liked to have seen a little more testing, and a comment on the algorithmic complexity.

You've taken the simple solution, but the assignment hints at a recursion-less solution. Can you see what that is?
